libparsec/crates/platform_storage/src/web/certificates.rs

```diff
@@ -1,28 +1,121 @@
-#![allow(unused_variables)]
 // Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 2016-present Scille SAS
 
-use std::{marker::PhantomData, path::Path};
+#![allow(unused_variables)]
+
+use std::collections::HashMap;
+use std::path::Path;
+use std::sync::Arc;
 
+use indexed_db_futures::prelude::IdbTransaction;
+use indexed_db_futures::IdbDatabase;
 use libparsec_types::prelude::*;
 
 use crate::certificates::{GetCertificateError, GetCertificateQuery, PerTopicLastTimestamps, UpTo};
+use crate::web::model::{Certificate, CertificateFilter};
+use crate::web::DB_VERSION;
+
+const COMMON_CERTIFICATES: [&str; 4] = [
+    "user_certificate",
+    "device_certificate",
+    "user_update_certificate",
+    "revoked_user_certificate",
+];
+
+const SEQUESTER_CERTIFICATES: [&str; 3] = [
+    "sequester_authority_certificate",
+    "sequester_service_certificate",
+    "sequester_revoked_service_certificate",
+];
+
+const REALM_CERTIFICATES: [&str; 4] = [
+    "realm_role_certificate",
+    "realm_name_certificate",
+    "realm_key_rotation_certificate",
+    "realm_archiving_certificate",
+];
+
+const SHAMIR_RECOVERY_CERTIFICATES: [&str; 2] = [
+    "shamir_recovery_share_certificate",
+    "shamir_recovery_brief_certificate",
+];
 
 #[derive(Debug)]
 pub(crate) struct PlatformCertificatesStorageForUpdateGuard<'a> {
-    _marker: PhantomData<&'a ()>,
+    transaction: IdbTransaction<'a>,
 }
 
 impl<'a> PlatformCertificatesStorageForUpdateGuard<'a> {
     pub async fn commit(self) -> anyhow::Result<()> {
-        todo!()
+        super::db::commit(self.transaction).await
     }
 
     pub async fn get_certificate_encrypted(
         &mut self,
         query: GetCertificateQuery,
         up_to: UpTo,
     ) -> Result<(DateTime, Vec<u8>), GetCertificateError> {
-        todo!()
+        // Handling `up_to` with a timestamp is a bit tricky:
+        // 1) we want the last result up to the given timestamp (included)
+        // 2) BUT ! if 1) fails we want the first result in order
+        //    to correctly return NonExisting vs ExistButTooRecent
+        //
+        // So to do that we act in three steps:
+        //  1) do the SQL query with the `up_to` filter
+        //  2) if the previous query returned nothing, do another SQL query without `up_to`
+        //  3) compare the datetime of the result with the on provided by `up_to`: if it's
+        //     too high it's a ExistButTooRecent error otherwise it's a valid result
+
+        // Step 1: do the regular SQL query
```

Comment:
The comment state about SQL, but we are currently using IDB + manual filtering / sorting

Reply:
Sorry

Reply:
Done